core/Pos/Binary/Core/Address.hs

```diff
@@ -1,95 +1,189 @@
+
+-- We don't need a special encoding for 'Address'', GND is what we want.
+deriving instance Bi Address'
+
+----------------------------------------------------------------------------
+-- Address serialization
+----------------------------------------------------------------------------
+
+{- NOTE: Address serialization
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
-----------------------------------------
+An address is serialized as a tuple consisting of:
 
-instance Bi (Attributes AddrPkAttrs) where
-    encode = encodeAttributes [(0, serialize' . addrPkDerivationPath)]
-    decode = decodeAttributes def $ \n v acc -> case n of
-        0 -> Just $ acc { addrPkDerivationPath = deserialize' v }
-        _ -> Nothing
+1. 'addrRoot'.
+2. 'addrAttributes'.
+3. 'addrType'.
+3. CRC32 checksum.
+-}
 
+instance CRC32 Address where
+    crc32Update seed =
+        crc32Update seed . CBOR.Write.toLazyByteString . encodeAddr
 
--- | Encodes the `Address` _without_ the CRC32.
+-- Encodes the `Address` __without__ the CRC32.
 -- It's important to keep this function separated from the `encode`
 -- definition to avoid that `encode` would call `crc32` and
 -- the latter invoke `crc32Update`, which would then try to call `encode`
 -- indirectly once again, in an infinite loop.
 encodeAddr :: Address -> Encoding
-encodeAddr addr = case addr of
-    PubKeyAddress keyHash attrs ->
-           encode (0 :: Word8)
-        <> encode (serialize' (keyHash, attrs))
-    ScriptAddress scrHash ->
-           encode (1 :: Word8)
-        <> encode (serialize' scrHash)
-    RedeemAddress keyHash ->
-           encode (2 :: Word8)
-        <> encode (serialize' keyHash)
-    UnknownAddressType t bs ->
-           encode t
-        <> encode bs
-
-instance Bi Address where
-    encode addr = encodeListLen 3 <> encodeAddr addr <> encode (crc32 addr)
+encodeAddr Address {..} =
+    encode addrRoot <> encode addrAttributes <> encode addrType
+
+-- Note: we are using 'Buildable' constraint here, which in turn
+-- relies on 'Bi', but it uses only encoding, while 'Buildable' is
+-- used here only in decoding.
+instance Buildable Address => Bi Address where
+    encode addr = encodeListLen 4 <> encodeAddr addr <> encode (crc32 addr)
     decode = do
-        enforceSize "Address" 3
-        t           <- decode @Word8
-        bs          <- decode @ByteString
-        expectedCRC <- decode @Word32
-        let address = case t of
-                          0 -> uncurry PubKeyAddress $ deserialize' bs
-                          1 -> ScriptAddress         $ deserialize' bs
-                          2 -> RedeemAddress         $ deserialize' bs
-                          _ -> UnknownAddressType t bs
-        let actualCRC = crc32 address
-        if expectedCRC /= actualCRC
-           then fail $ toString $ format ("Address " % shown % "has invalid checksum: " % shown) address actualCRC
-           else pure address
+        (addrRoot, addrAttributes, addrType, decodedCRC) <- decode
```

Comment:
will it work? didn't you forget length? anyway, it will be better to check `enforceSize`